[Bug]: Warnings when compiled on FreeBSD (plugins)

### Bug description

When we compile Netdata on `FreeBSD` the following warnings are shown:

```sh
[/home/thiago/Netdata/removeme]# gmake -j2 
collectors/apps.plugin/apps_plugin.c: In function 'read_global_time':
collectors/apps.plugin/apps_plugin.c:1643:56: warning: unused variable 'gtime_raw' [-Wunused-variable]
 1643 |     static kernel_uint_t utime_raw = 0, stime_raw = 0, gtime_raw = 0, ntime_raw = 0;
      |                                                        ^~~~~~~~~
In file included from ./aclk/aclk_proxy.h:4,
                 from ./aclk/aclk_api.h:7,
                 from ./daemon/common.h:72,
                 from ./collectors/all.h:6,
                 from collectors/apps.plugin/apps_plugin.c:9:
collectors/apps.plugin/apps_plugin.c: In function 'read_pid_file_descriptors':
collectors/apps.plugin/apps_plugin.c:2001:27: warning: comparison of integer expressions of different signedness: 'int' and 'size_t' {aka 'long unsigned int'} [-Wsign-compare]
 2001 |         if (unlikely(fdid >= p->fds_size)) {
      |                           ^~
./config.h:404:41: note: in definition of macro 'unlikely'
  404 | #define unlikely(x) __builtin_expect(!!(x), 0)
      |                                         ^
At top level:
collectors/apps.plugin/apps_plugin.c:1643:56: warning: 'gtime_raw' defined but not used [-Wunused-variable]
 1643 |     static kernel_uint_t utime_raw = 0, stime_raw = 0, gtime_raw = 0, ntime_raw = 0;
      |                                                        ^~~~~~~~~
collectors/apps.plugin/apps_plugin.c:3852:13: warning: 'send_proc_states_count' defined but not used [-Wunused-function]
 3852 | static void send_proc_states_count(usec_t dt)
      |             ^~~~~~~~~~~~~~~~~~~~~~
collectors/apps.plugin/apps_plugin.c: In function 'read_pid_file_descriptors':
collectors/apps.plugin/apps_plugin.c:2058:51: warning: '%s' directive writing up to 1023 bytes into a region of size 1013 [-Wformat-overflow=]
 2058 |                     sprintf(fdsname, "other: shm: %s size: %lu", fds->kf_path, fds->kf_un.kf_file.kf_file_size);
      |                                                   ^~
collectors/apps.plugin/apps_plugin.c:2058:21: note: 'sprintf' output between 21 and 1063 bytes into a destination of size 1025
 2058 |                     sprintf(fdsname, "other: shm: %s size: %lu", fds->kf_path, fds->kf_un.kf_file.kf_file_size);
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
libnetdata/libnetdata.c: In function 'madvise_mergeable':
libnetdata/libnetdata.c:988:43: warning: unused parameter 'mem' [-Wunused-parameter]
  988 | static inline int madvise_mergeable(void *mem, size_t len) {
      |                                     ~~~~~~^~~
libnetdata/libnetdata.c:988:55: warning: unused parameter 'len' [-Wunused-parameter]
  988 | static inline int madvise_mergeable(void *mem, size_t len) {
      |                                                ~~~~~~~^~~
collectors/freebsd.plugin/freebsd_sysctl.c: In function 'do_system_ram':
collectors/freebsd.plugin/freebsd_sysctl.c:1004:24: warning: unused variable 'rd_cache' [-Wunused-variable]
 1004 |                       *rd_cache = NULL, *rd_buffers = NULL;
      |                        ^~~~~~~~
collectors/freebsd.plugin/freebsd_sysctl.c:976:16: warning: unused variable 'mib_cache_count' [-Wunused-variable]
  976 |                mib_cache_count[4] = {0, 0, 0, 0}, mib_vfs_bufspace[2] = {0, 0}, mib_free_count[4] = {0, 0, 0, 0};
      |                ^~~~~~~~~~~~~~~
At top level:
collectors/freebsd.plugin/freebsd_sysctl.c:976:16: warning: 'mib_cache_count' defined but not used [-Wunused-variable]
collectors/freebsd.plugin/freebsd_sysctl.c:1004:24: warning: 'rd_cache' defined but not used [-Wunused-variable]
 1004 |                       *rd_cache = NULL, *rd_buffers = NULL;
      |                        ^~~~~~~~
collectors/freebsd.plugin/freebsd_ipfw.c: In function 'do_ipfw':
collectors/freebsd.plugin/freebsd_ipfw.c:236:45: warning: format '%d' expects argument of type 'int', but argument 3 has type 'uint32_t' {aka 'unsigned int'} [-Wformat=]
  236 |                     sprintf(rule_num_str, "%d_%d", rule->rulenum, rule->id);
      |                                            ~^      ~~~~~~~~~~~~~
      |                                             |          |
      |                                             int        uint32_t {aka unsigned int}
      |                                            %d
collectors/freebsd.plugin/freebsd_ipfw.c:236:48: warning: format '%d' expects argument of type 'int', but argument 4 has type 'uint32_t' {aka 'unsigned int'} [-Wformat=]
  236 |                     sprintf(rule_num_str, "%d_%d", rule->rulenum, rule->id);
      |                                               ~^                  ~~~~~~~~
      |                                                |                      |
      |                                                int                    uint32_t {aka unsigned int}
      |                                               %d
 OK  ''

```

### Expected behavior

Compile Netdata without warnings.

### Steps to reproduce

1. Clone Netdata
2. Compile `master` branch using the flags `CFLAGS="-O1 -ggdb -Wall -Wextra -fno-omit-frame-pointer -Wformat-signedness -fstack-protector-all -DNETDATA_INTERNAL_CHECKS=1 -D_FORTIFY_SOURCE=2 -DNETDATA_VERIFY_LOCKS=1"`

...


### Installation method

from source

### System info

```shell
FreeBSD freebsd 13.0-RELEASE-p11 FreeBSD 13.0-RELEASE-p11 #0: Tue Apr  5 18:54:35 UTC 2022     root@amd64-builder.daemonology.net:/usr/obj/usr/src/amd64.amd64/sys/GENERIC  amd64
```


### Netdata build info

```shell
Version: netdata v1.34.0-69-nightly
Configure options:  '--prefix=/usr' '--sysconfdir=/etc' '--localstatedir=/var' '--libexecdir=/usr/libexec' '--libdir=/usr/lib' '--with-zlib' '--with-math' '--with-user=netdata' '--disable-lto' '--disable-ml' '--disable-dependency-tracking' '--with-bundled-protobuf' '--with-bundled-libJudy' 'CFLAGS=-O1 -ggdb -Wall -Wextra -fno-omit-frame-pointer -Wformat-signedness -fstack-protector-all -DNETDATA_INTERNAL_CHECKS=1 -D_FORTIFY_SOURCE=2 -DNETDATA_VERIFY_LOCKS=1' 'LDFLAGS='
Install type: custom
Features:
    dbengine:                   YES
    Native HTTPS:               YES
    Netdata Cloud:              YES 
    ACLK Next Generation:       YES
    ACLK-NG New Cloud Protocol: YES
    ACLK Legacy:                NO
    TLS Host Verification:      YES
    Machine Learning:           NO
    Stream Compression:         YES
Libraries:
    protobuf:                YES (bundled)
    jemalloc:                NO
    JSON-C:                  YES
    libcap:                  NO
    libcrypto:               YES
    libm:                    YES
    tcalloc:                 NO
    zlib:                    YES
Plugins:
    apps:                    YES
    cgroup Network Tracking: NO
    CUPS:                    NO
    EBPF:                    NO
    IPMI:                    NO
    NFACCT:                  NO
    perf:                    NO
    slabinfo:                NO
    Xen:                     NO
    Xen VBD Error Tracking:  NO
Exporters:
    AWS Kinesis:             NO
    GCP PubSub:              NO
    MongoDB:                 NO
    Prometheus Remote Write: NO
```


### Additional info

_No response_

--- a/collectors/apps.plugin/apps_plugin.c
+++ b/collectors/apps.plugin/apps_plugin.c
@@ -118,6 +118,7 @@ typedef enum {
     PROC_STATUS_END, //place holder for ending enum fields
 } proc_state;
 
+#ifndef __FreeBSD__
 static proc_state proc_state_count[PROC_STATUS_END];
 static const char *proc_states[] = {
     [PROC_STATUS_RUNNING] = "running",
@@ -126,6 +127,7 @@ static const char *proc_states[] = {
     [PROC_STATUS_ZOMBIE] = "zombie",
     [PROC_STATUS_STOPPED] = "stopped",
     };
+#endif
 
 // ----------------------------------------------------------------------------
 // internal flags
@@ -1252,7 +1254,6 @@ void arl_callback_status_rssshmem(const char *name, uint32_t hash, const char *v
 
     aptr->p->status_rssshmem = str2kernel_uint_t(procfile_lineword(aptr->ff, aptr->line, 1));
 }
-#endif // !__FreeBSD__
 
 static void update_proc_state_count(char proc_state) {
     switch (proc_state) {
@@ -1275,6 +1276,7 @@ static void update_proc_state_count(char proc_state) {
             break;
     }
 }
+#endif // !__FreeBSD__
 
 static inline int read_proc_pid_status(struct pid_stat *p, void *ptr) {
     p->status_vmsize           = 0;
@@ -1495,7 +1497,9 @@ static inline int read_proc_pid_stat(struct pid_stat *p, void *ptr) {
         p->cstime           = 0;
         p->cgtime           = 0;
     }
+#ifndef __FreeBSD__
     update_proc_state_count(p->state);
+#endif
     return 1;
 
 cleanup:
@@ -1640,7 +1644,7 @@ cleanup:
 }
 #else
 static inline int read_global_time() {
-    static kernel_uint_t utime_raw = 0, stime_raw = 0, gtime_raw = 0, ntime_raw = 0;
+    static kernel_uint_t utime_raw = 0, stime_raw = 0, ntime_raw = 0;
     static usec_t collected_usec = 0, last_collected_usec = 0;
     long cp_time[CPUSTATES];
 
@@ -1995,7 +1999,7 @@ static inline int read_pid_file_descriptors(struct pid_stat *p, void *ptr) {
         }
 
         // get file descriptors array index
-        int fdid = fds->kf_fd;
+        size_t fdid = fds->kf_fd;
 
         // check if the fds array is small
         if (unlikely(fdid >= p->fds_size)) {
@@ -2055,7 +2059,12 @@ static inline int read_pid_file_descriptors(struct pid_stat *p, void *ptr) {
 #endif
                     break;
                 case KF_TYPE_SHM:
-                    sprintf(fdsname, "other: shm: %s size: %lu", fds->kf_path, fds->kf_un.kf_file.kf_file_size);
+                    sprintf(
+                        fdsname,
+                        "other: shm: %s size: %lu",
+                        fds->kf_path,
+                        (FILENAME_MAX < fds->kf_un.kf_file.kf_file_size) ? FILENAME_MAX :
+                                                                           fds->kf_un.kf_file.kf_file_size);
                     break;
                 case KF_TYPE_SEM:
                     sprintf(fdsname, "other: sem: %u", fds->kf_un.kf_sem.kf_sem_value);
@@ -2575,9 +2584,10 @@ static inline int collect_data_for_pid(pid_t pid, void *ptr) {
 static int collect_data_for_all_processes(void) {
     struct pid_stat *p = NULL;
 
+#ifndef __FreeBSD__
     // clear process state counter
     memset(proc_state_count, 0, sizeof proc_state_count);
-#ifdef __FreeBSD__
+#else
     int i, procnum;
 
     static size_t procbase_size = 0;
@@ -3849,6 +3859,8 @@ static void send_charts_updates_to_netdata(struct target *root, const char *type
     }
 }
 
+
+#ifndef __FreeBSD__
 static void send_proc_states_count(usec_t dt)
 {
     static bool chart_added = false;
@@ -3872,6 +3884,7 @@ static void send_proc_states_count(usec_t dt)
     }
     send_END();
 }
+#endif
 
 // ----------------------------------------------------------------------------
 // parse command line arguments

--- a/collectors/apps.plugin/apps_plugin.c
+++ b/collectors/apps.plugin/apps_plugin.c
@@ -1962,6 +1962,8 @@ static inline int read_pid_file_descriptors(struct pid_stat *p, void *ptr) {
     static char *fdsbuf;
     char *bfdsbuf, *efdsbuf;
     char fdsname[FILENAME_MAX + 1];
+#define SHM_FORMAT_LEN 31 // format: 21 + size: 10
+    char shm_name[FILENAME_MAX - SHM_FORMAT_LEN + 1];
 
     // we make all pid fds negative, so that
     // we can detect unused file descriptors
@@ -2059,12 +2061,8 @@ static inline int read_pid_file_descriptors(struct pid_stat *p, void *ptr) {
 #endif
                     break;
                 case KF_TYPE_SHM:
-                    sprintf(
-                        fdsname,
-                        "other: shm: %s size: %lu",
-                        fds->kf_path,
-                        (FILENAME_MAX < fds->kf_un.kf_file.kf_file_size) ? FILENAME_MAX :
-                                                                           fds->kf_un.kf_file.kf_file_size);
+                    strncpyz(shm_name, fds->kf_path, FILENAME_MAX - SHM_FORMAT_LEN);
+                    sprintf(fdsname, "other: shm: %s size: %lu", shm_name, fds->kf_un.kf_file.kf_file_size);
                     break;
                 case KF_TYPE_SEM:
                     sprintf(fdsname, "other: sem: %u", fds->kf_un.kf_sem.kf_sem_value);

--- a/collectors/freebsd.plugin/freebsd_ipfw.c
+++ b/collectors/freebsd.plugin/freebsd_ipfw.c
@@ -233,7 +233,7 @@ int do_ipfw(int update_every, usec_t dt) {
                     break;
 
                 if (likely(do_static)) {
-                    sprintf(rule_num_str, "%d_%d", rule->rulenum, rule->id);
+                    sprintf(rule_num_str, "%"PRIu32"_%d", (uint32_t)rule->rulenum, rule->id);
 
                     rd_packets = rrddim_find_active(st_packets, rule_num_str);
                     if (unlikely(!rd_packets))

--- a/collectors/freebsd.plugin/freebsd_ipfw.c
+++ b/collectors/freebsd.plugin/freebsd_ipfw.c
@@ -233,7 +233,7 @@ int do_ipfw(int update_every, usec_t dt) {
                     break;
 
                 if (likely(do_static)) {
-                    sprintf(rule_num_str, "%"PRIu32"_%d", (uint32_t)rule->rulenum, rule->id);
+                    sprintf(rule_num_str, "%"PRIu32"_%"PRIu32"", (uint32_t)rule->rulenum, (uint32_t)rule->id);
 
                     rd_packets = rrddim_find_active(st_packets, rule_num_str);
                     if (unlikely(!rd_packets))

--- a/collectors/freebsd.plugin/freebsd_sysctl.c
+++ b/collectors/freebsd.plugin/freebsd_sysctl.c
@@ -972,8 +972,14 @@ int do_vm_swap_info(int update_every, usec_t dt) {
 
 int do_system_ram(int update_every, usec_t dt) {
     (void)dt;
-    static int mib_active_count[4] = {0, 0, 0, 0}, mib_inactive_count[4] = {0, 0, 0, 0}, mib_wire_count[4] = {0, 0, 0, 0},
-               mib_cache_count[4] = {0, 0, 0, 0}, mib_vfs_bufspace[2] = {0, 0}, mib_free_count[4] = {0, 0, 0, 0};
+    static int mib_active_count[4] = {0, 0, 0, 0},
+               mib_inactive_count[4] = {0, 0, 0, 0},
+               mib_wire_count[4] = {0, 0, 0, 0},
+#if __FreeBSD_version < 1200016
+               mib_cache_count[4] = {0, 0, 0, 0},
+#endif
+               mib_vfs_bufspace[2] = {0, 0},
+               mib_free_count[4] = {0, 0, 0, 0};
     vmmeter_t vmmeter_data;
     size_t vfs_bufspace_count;
 

--- a/libnetdata/libnetdata.c
+++ b/libnetdata/libnetdata.c
@@ -982,6 +982,9 @@ static inline int madvise_mergeable(void *mem, size_t len) {
     if (ret != 0 && logger-- > 0) error("madvise(MADV_MERGEABLE) failed.");
     return ret;
 #else
+    UNUSED(mem);
+    UNUSED(len);
+    
     return 0;
 #endif
 }

